libdnf/sack/Solution.cpp

```diff
@@ -23,61 +23,56 @@
 
 namespace libdnf {
 
-bool
-Solution::getBestSolution(const char * subject, DnfSack* sack, HyForm * forms, bool icase,
-    bool with_nevra, bool with_provides, bool with_filenames, bool with_src)
+Solution::Solution(const char * subject, DnfSack* sack, HyForm * forms, bool icase,
+    bool with_nevra, bool with_provides, bool with_filenames, bool with_src,
+    Query::ExcludeFlags exclude_flags)
+:
+    query(new Query(sack, exclude_flags))
 {
-    nevra.reset();
-    Query baseQuery(sack);
+    Query baseQuery(*query);
     if (!with_src) {
         baseQuery.addFilter(HY_PKG_ARCH, HY_NEQ, "src");
     }
-    baseQuery.apply();
-    std::unique_ptr<Query> queryCandidate(new Query(baseQuery));
+
+    query->addFilter(HY_PKG_EMPTY, HY_EQ, 1);
+
     if (with_nevra) {
         Nevra nevraObj;
         const HyForm * tryForms = !forms ? HY_FORMS_MOST_SPEC : forms;
         for (std::size_t i = 0; tryForms[i] != _HY_FORM_STOP_; ++i) {
             if (nevraObj.parse(subject, tryForms[i])) {
-                queryCandidate->queryUnion(baseQuery);
-                queryCandidate->addFilter(&nevraObj, icase);
-                if (!queryCandidate->empty()) {
+                query->queryUnion(baseQuery);
+                query->addFilter(&nevraObj, icase);
+                if (!query->empty()) {
                     nevra.reset(new Nevra(std::move(nevraObj)));
-                    query = std::move(queryCandidate);
-                    return true;
```

Comment:
The return value was a performance improvement, because query.empty() is relatively costly request.

Reply:
Oh, I see, I haven't gone that deep. Still, the value wasn't used in the current code... probably better to keep it simple until it is needed?

Reply:
The code of _get_best_selector that is in development already use it or will use it. Please we have to think about design in advance. Or do you want to overwrite the code again and again?

Reply:
Where is the code? I didn't know about it. The design of the class seemed ad-hoc to me...

Reply:
https://github.com/rpm-software-management/libdnf/pull/472/files
in goal.cpp

Reply:
The solution class was created as a part of the PR, but it cannot be finished due to compatibility for dnf4.

Reply:
I see, thanks. I still think my changes make for a better design of the class. To solve the `empty()` issue, I would cache the emptiness value somewhere on the line of `Solution` - `Query` - `PackageSet`. The lower the better, but since both `Query` and `PackageSet` allow to access the underlying `Map` struct, we'd probably have to cache it in `Solution`. Ideally we get to a state when these classes don't need to leak their internal state and we can implement this optimization on the `PackageSet` level.

> Please we have to think about design in advance. Or do you want to overwrite the code again and again?

Iterating is inevitable to achieve quality code, especially if we want to turn it into a public API at some point. But I do agree thinking about the design carefully in advance saves the number of iterations...

Reply:
I've removed the commit that changed the `getBestSolution()` method to constructor.